tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java

```diff
@@ -0,0 +1,397 @@
+                        if (line == null) {
+                            throw new IOException("EOF");
+                        }
+                    } catch (IOException e) {
+                        log.info("{}: can't read any more from stdout: {}", id, e.getMessage());
+                        return;
+                    }
+                    log.trace("{}: read line from stdin: {}", id, line);
+                    JsonNode resp = readObject(line);
+                    if (resp.has("status")) {
+                        log.info("{}: New status: {}", id, resp.get("status").toString());
+                        status.update(resp.get("status"));
+                    }
+                    if (resp.has("log")) {
+                        log.info("{}: (stdout): {}", id, resp.get("log").asText());
+                    }
+                    if (resp.has("error")) {
+                        String error = resp.get("error").asText();
+                        log.error("{}: error: {}", id, error);
+                        doneFuture.complete(error);
+                    }
+                }
+            } catch (Throwable e) {
+                log.info("{}: error reading from stdout.", id, e);
+            }
+        }
+    }
+
+    class StderrMonitor implements Runnable {
+        private final Process process;
+
+        StderrMonitor(Process process) {
+            this.process = process;
+        }
+
+        @Override
+        public void run() {
+            log.trace("{}: starting stderr monitor.", id);
+            try (BufferedReader br = new BufferedReader(
+                new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8))) {
+                String line;
+                while (true) {
+                    try {
+                        line = br.readLine();
+                        if (line == null) {
+                            throw new IOException("EOF");
+                        }
+                    } catch (IOException e) {
+                        log.info("{}: can't read any more from stderr: {}", id, e.getMessage());
+                        return;
+                    }
+                    log.error("{}: (stderr):{}", id, line);
+                }
+            } catch (Throwable e) {
+                log.info("{}: error reading from stderr.", id, e);
+            }
+        }
+    }
+
+    class StdinWriter implements Runnable {
+        private final Process process;
+
+        StdinWriter(Process process) {
+            this.process = process;
+        }
+
+        @Override
+        public void run() {
+            OutputStreamWriter stdinWriter = new OutputStreamWriter(
+                process.getOutputStream(), StandardCharsets.UTF_8);
+            try {
+                while (true) {
+                    log.info("{}: stdin writer ready.", id);
+                    JsonNode node = stdinQueue.take();
+                    if (node instanceof NullNode) {
```

Comment:
We using NullNode as a sentinel value here?

Reply:
I guess I can just use an `Optional<JsonNode>` to make it clearer.